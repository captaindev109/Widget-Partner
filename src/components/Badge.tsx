import { FC } from 'react';
import styled from 'styled-components';
import Card from '@/generics/Card';

export interface BadgeProps {
  children: string;
}

const StyledBadge = styled(Card)<BadgeProps>`
  background-color: ${(props) => props.theme.color.purple300};
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  border-radius: 4px;
  padding: 5px 8px 3px;
  line-height: 16px;
  font-size: 12px;
  color: ${(props) => props.theme.color.gray900};
`;

const Badge: FC<BadgeProps> = ({ children }) => {
  return <StyledBadge data-testid="badge">{children.toUpperCase()}</StyledBadge>;
};

const defaultProps: Partial<BadgeProps> = {};

Badge.defaultProps = defaultProps;

export default Badge;
