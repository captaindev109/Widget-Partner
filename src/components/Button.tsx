import styled, { DefaultTheme } from 'styled-components';
import { FC } from 'react';
import type { BasicBool, ButtonProps, ButtonVariant } from '@/types/types';
import Card from '@/generics/Card';
import { CreateSVGComponent } from './SVGIcons/Index';

interface VariantSpecificStyling {
  variant?: 'primary' | 'outline' | 'transparent' | 'icon';
  fluid?: boolean;
  theme: DefaultTheme;
}

const getVariantSpecificStyling = ({ theme, variant }: VariantSpecificStyling) => {
  switch (variant) {
    case 'primary':
      return `
        cursor: pointer;
        background-color: ${theme.color.gray900};
        color: ${theme.color.gray100};
        border:none
        cursor: pointer;
        &:hover {
          cursor: pointer;
          background-color: ${theme.color.gray800};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.color.gray300};
          color: ${theme.color.gray500};
        }
      `;
    case 'outline':
      return `
        background-color: ${theme.palette.common.transparent};
        color: ${theme.color.gray900};
        border: 1px solid ${theme.color.gray900};
        &:hover {
          cursor: pointer;
          border: 1px solid ${theme.color.gray700};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.palette.common.transparent};
          border: 1px solid ${theme.color.gray500};
          color: ${theme.color.gray500};
        }
      `;
    case 'transparent':
      return `
        background-color: ${theme.palette.common.transparent};
        color: ${theme.color.gray900};
        &:hover {
          cursor: pointer;
          background-color: ${theme.color.gray300};
        }
        &:disabled {
          cursor: default;
          background-color: ${theme.palette.common.transparent};
          color: ${theme.color.gray500};
          & > div > div > svg {
            color: ${theme.color.gray500};
            fill: ${theme.color.gray500};
          }
        }
      `;
    case 'icon':
      return `
        background: none;
        cursor: pointer;
        padding: 0;
        height: 32px;
        width: 32px;
        z-index: 1;
      `;
    default:
      return ``;
  }
};

const getIconColor = ({
  variant,
  disabled,
  theme,
}: {
  variant?: ButtonVariant;
  disabled?: BasicBool;
  theme: DefaultTheme;
}) => {
  if (disabled) return theme.color.gray500;
  if (variant === 'primary') return theme.color.gray100;
  return theme.color.gray900;
};

const StyledButton = styled.button<VariantSpecificStyling>`
  display: flex;
  flex: 1;
  width: ${(props) => (props.fluid ? 'auto' : '100%')};
  align-items: center;
  justify-content: center;
  padding: 0px 32px;
  margin: 0;
  height: 48px;
  border-radius: 8px;
  border: none;
  pointer: cursor;
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  ${getVariantSpecificStyling}
`;

const ContentWrapper = styled(Card)`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const IconWrapper = styled(Card)<{ disabled?: BasicBool; variant?: ButtonVariant }>`
  ${(props) => props.variant !== 'icon' && `padding-left: 16px;`}
  display: flex;
  align-items: center;
  justify-content: center;
  & > svg > * {
    fill: ${getIconColor};
  }
`;

const Button: FC<ButtonProps> = ({ children, onClick, icon, ...rest }) => {
  return (
    <StyledButton onClick={onClick} {...rest}>
      <ContentWrapper>
        {children}
        {icon && (
          <IconWrapper variant={rest.variant} disabled={rest.disabled}>
            {CreateSVGComponent(icon.iconName, icon.config)}
          </IconWrapper>
        )}
      </ContentWrapper>
    </StyledButton>
  );
};

Button.defaultProps = {
  type: 'button',
  disabled: false,
  variant: 'primary',
};

export default Button;
