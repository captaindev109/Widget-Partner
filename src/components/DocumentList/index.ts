export { DocumentLink, DocumentList } from './DocumentList';
