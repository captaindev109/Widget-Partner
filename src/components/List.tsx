import { ReactNode, FC } from 'react';
import styled from 'styled-components';

export interface ListProps {
  children: ReactNode;
  display?: 'flex' | 'block' | 'inline';
  direction?: 'row' | 'column';
  spaceBetween?: string;
  wrap?: 'wrap' | 'nowrap';
  className?: string;
}

export interface ListItemProps {
  children: ReactNode;
  className?: string;
}

const StyledList = styled.ul<ListProps>`
  display: ${(props) => props.display};
  flex-direction: ${(props) => props.direction};
  flex-wrap: ${(props) => props.wrap};
  list-style: none;

  & > * {
    ${(props) =>
      props.direction === 'row' && props.spaceBetween
        ? `margin-left: ${props.spaceBetween};`
        : `margin-top: ${props.spaceBetween};`}
  }

  & > *:first-child {
    ${(props) => (props.direction === 'row' && props.spaceBetween ? 'margin-left: 0px;' : 'margin-top: 0px;')}
  }
`;

export const StyledListItem = styled.li<ListItemProps>``;

export const List: FC<ListProps> = ({ children, display, direction, spaceBetween, wrap, className }): JSX.Element => {
  return (
    <StyledList className={className} wrap={wrap} spaceBetween={spaceBetween} display={display} direction={direction}>
      {children}
    </StyledList>
  );
};

export const ListItem: FC<ListItemProps> = ({ children, className }): JSX.Element => {
  return <StyledListItem className={className}>{children}</StyledListItem>;
};
