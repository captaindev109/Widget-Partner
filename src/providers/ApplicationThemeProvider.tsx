import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
