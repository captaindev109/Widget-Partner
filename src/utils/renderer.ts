import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
