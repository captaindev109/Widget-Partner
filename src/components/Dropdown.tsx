import Card from '@/generics/Card';
import Label from '@/generics/Label';
import styled from 'styled-components';
import Tooltip from './Tooltip';

import { CreateSVGComponent } from '@/components/SVGIcons/Index';
import Text from '@/generics/Text';
import { animateOpacity, ANIMATION_DURATION } from '@/utils/animationUtils';
import { BasicString, BasicBool, BasicAny } from '@/types/types';
import { FC, useEffect, useRef } from 'react';

export interface DropdownOption {
  label: string;
  value: string;
}

export interface DropdownProps {
  options: DropdownOption[];
  onChange?: (option: DropdownOption) => void;
  value?: BasicString;
  disabled?: BasicBool;
  placeholder?: BasicString;
  label?: BasicString;
  name?: BasicString;
  tooltip?: BasicString;
  helperText?: BasicString;
  error?: BasicBool;
  icon?: BasicBool;
  register?: BasicAny;
}

export interface DropdownOption {
  label: string;
  value: string;
}

const TopRow = styled(Card)`
  display: flex;
  align-items: center;

  & > * {
    margin-left: 10px;
  }

  margin-bottom: 8px;
`;

const StyledIcon = styled(Card)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: all ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const HelperText = styled(Text)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: color ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const IconWrapper = styled(Card)`
  position: absolute;
  top: 0;
  bottom: 0;
  right: 24px;

  display: flex;
  align-items: center;
  pointer-events: none;
`;

const StyledSelect = styled.select<BasicAny>`
  outline: none;
  appearance: none;
  background: white;
  border: none;
  display: flex;

  width: 100%;
  margin: 0;
  height: 58px;
  padding: 16px 24px;
  border-radius: ${({ theme }) => theme.size.borderRadius.br4};
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  font-size: 16px;
  line-height: 24px;
  color: ${(props) => props.theme.color.gray900};
  border: 1px solid ${(props) => props.theme.color.gray300};

  transition: border-color 140ms ease-in;

  &:selected + ${IconWrapper} {
    border: 1px solid ${(props) => props.theme.color.gray900};
  }

  &:hover {
    cursor: pointer;
    border: 1px solid ${(props) => props.theme.color.gray900};
  }

  &:disabled {
    cursor: default;
    background-color: ${(props) => props.theme.color.gray300};
    border: 1px solid ${(props) => props.theme.color.gray300};
    &:hover {
      cursor: default;
    }
  }
`;

const Dropdown: FC<DropdownProps> = ({
  tooltip,
  label,
  options,
  disabled,
  helperText,
  icon,
  register,
  error,
  ...rest
}) => {
  const didMountRef = useRef<boolean>(false);

  useEffect(() => {
    didMountRef.current = true;
  }, []);

  return (
    <Card>
      <TopRow>
        {label && <Label data-testid="dropdown-label">{label}</Label>}
        {!!tooltip && <Tooltip direction="top" text={tooltip} size="small"></Tooltip>}
      </TopRow>
      <Card position="relative">
        <StyledSelect disabled={disabled} {...register} {...rest}>
          {options.map((option) => (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          ))}
        </StyledSelect>
        <IconWrapper>{CreateSVGComponent('DownArrow')}</IconWrapper>
      </Card>

      {!!helperText && (
        <Card display="flex" alignItems="center" marginRight="4px">
          {icon && (
            <StyledIcon error={error} $animate={didMountRef.current} marginRight="4px">
              {CreateSVGComponent(error ? 'Error' : 'HelperInfo')}
            </StyledIcon>
          )}
          <HelperText data-testid="text-helper" error={error} $animate={didMountRef.current}>
            {helperText}
          </HelperText>
        </Card>
      )}
    </Card>
  );
};

export default Dropdown;
