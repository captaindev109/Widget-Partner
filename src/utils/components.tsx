import Page from '@/generics/Page';
import Card from '@/generics/Card';
import Image from '@/generics/Image';
import Title from '@/generics/Title';
import Footer from '@/components/Footer';
import Button from '@/components/Button';
import RadioGroup from '@/components/RadioGroup';
import IconButton from '@/components/IconButton';
import RedirectButton from '@/components/RedirectButton';
import PolicyText from '@/components/PolicyText';
import GetStarted from '@/components/GetStarted';
import Main from '@/containers/Main';
import Header from '@/containers/Header';
import MainTopContainer from '@/containers/MainTopContainer';
import LinkText from '@/generics/LinkText';
import Text from '@/generics/Text';
import Input from '@/generics/Input';
import Wrapper from '@/generics/Wrapper';
import { List, ListItem } from '@/components/List';
import IconWithLabel from '@/components/IconWithLabel';
import MainBottomContainer from '@/containers/MainBottomContainer';
import HeaderContainer from '@/containers/HeaderContainer/HeaderContainer';
import { DocumentLink, DocumentList } from '@/components/DocumentList';
import { InformationList, InformationItem } from '@/components/InformationList';
import { Tab, Tabs } from '@/components/Tabs';
import ToggleSwitch from '@/generics/ToggleSwitch';
import Span from '@/generics/Span';
import FooterWithQuote from '@/components/FooterWithQuote';
import DynamicForm from '@/components/DynamicForm';
import Dropdown from '@/components/Dropdown';
import HeaderWithImage from '@/components/HeaderWithImage';
import HeaderWithNavigation from '@/components/HeaderWithNavigation';
import NavigationButton from '@/components/NavigationButton';
import ExoticOfferDocuments from '@/components/DocumentList/ExoticOfferDocuments';
import ExoticCheckoutDocuments from '@/components/DocumentList/ExoticCheckoutDocuments';

import { BasicAny, BasicString } from '@/types/types';
import ExoticTabs from '@/components/Tabs/ExoticTabs';
import ExoticPriceOffer from '@/components/PriceOffer/ExoticPriceOffer';
import ExoticInformationList from '@/components/InformationList/ExoticInformationList';
import ExoticProductList from '@/components/ProductList/ExoticProductList';
import ExoticInsuranceSummary from '@/components/InsuranceSummary/ExoticInsuranceSummary';
import ExoticConfirmationText from '@/components/ConfirmationText/ExoticConfirmationText';
import InsuranceSummary from '@/components/InsuranceSummary';
import ExoticStartDate from '@/components/StartDate/ExoticStartDate';
import TrustlyIframe from '@/components/TrustlyIframe';
import ExoticTrustlyIframe from '@/components/TrustlyIframe/ExoticTrustlyIframe';
import Spinner from '@/components/Spinner';

export const UIComponents: Record<BasicString, BasicAny> = {
  Page: Page,
  Card: Card,
  Image: Image,
  Title: Title,
  Text: Text,
  LinkText: LinkText,
  Header: Header,
  Main: Main,
  Footer: Footer,
  Button: Button,
  Input: Input,
  RadioGroup: RadioGroup,
  RedirectButton: RedirectButton,
  IconButton: IconButton,
  PolicyText: PolicyText,
  GetStarted: GetStarted,
  Wrapper: Wrapper,
  List: List,
  ListItem: ListItem,
  IconWithLabel: IconWithLabel,
  MainTopContainer: MainTopContainer,
  MainBottomContainer: MainBottomContainer,
  HeaderContainer: HeaderContainer,
  DocumentLink: DocumentLink,
  DocumentList: DocumentList,
  FooterWithQuote: FooterWithQuote,
  InformationList: InformationList,
  InformationItem: InformationItem,
  Tab: Tab,
  Tabs: Tabs,
  ToggleSwitch: ToggleSwitch,
  Span: Span,
  DynamicForm: DynamicForm,
  Dropdown: Dropdown,
  HeaderWithImage: HeaderWithImage,
  HeaderWithNavigation: HeaderWithNavigation,
  NavigationButton: NavigationButton,
  ExoticOfferDocuments: ExoticOfferDocuments,
  ExoticCheckoutDocuments: ExoticCheckoutDocuments,
  ExoticTabs: ExoticTabs,
  ExoticPriceOffer: ExoticPriceOffer,
  ExoticInformationList: ExoticInformationList,
  ExoticProductList: ExoticProductList,
  ExoticInsuranceSummary: ExoticInsuranceSummary,
  ExoticStartDate: ExoticStartDate,
  ExoticConfirmationText: ExoticConfirmationText,
  InsuranceSummary: InsuranceSummary,
  TrustlyIframe: TrustlyIframe,
  ExoticTrustlyIframe: ExoticTrustlyIframe,
  Spinner: Spinner,
};
