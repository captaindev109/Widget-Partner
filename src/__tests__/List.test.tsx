import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import defaultTheme from '@/theme/theme';
import { List, ListItem } from '@/components/List';

const MockComponent: React.FC = () => {
  return (
    <ApplicationThemeProvider theme={defaultTheme}>
      <List>
        <ListItem>
          <p>text 1</p>
        </ListItem>
        <ListItem>
          <p>text 2</p>
        </ListItem>
        <ListItem>
          <p>text 3</p>
        </ListItem>
        <ListItem>
          <p>text 4</p>
        </ListItem>
      </List>
    </ApplicationThemeProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<List /> Test Cases', () => {
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponent />);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
