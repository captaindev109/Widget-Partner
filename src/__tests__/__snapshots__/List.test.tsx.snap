// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> Test Cases Snapshot test should pass 1`] = `
<ul
  className="List__StyledList-sc-1b1or9q-0 iyrMoT"
>
  <li
    className="List__StyledListItem-sc-1b1or9q-1"
  >
    <p>
      text 1
    </p>
  </li>
  <li
    className="List__StyledListItem-sc-1b1or9q-1"
  >
    <p>
      text 2
    </p>
  </li>
  <li
    className="List__StyledListItem-sc-1b1or9q-1"
  >
    <p>
      text 3
    </p>
  </li>
  <li
    className="List__StyledListItem-sc-1b1or9q-1"
  >
    <p>
      text 4
    </p>
  </li>
</ul>
`;
